braille-printer-app/generic-brf.c

```diff
@@ -0,0 +1,237 @@
+    pappl_device_t *device)      // I - Output device
+{
+  int fd;             // Input file
+  ssize_t bytes;      // Bytes read/written
+  char buffer[10000007]; // Read/write buffer
+
+  // Copy the raw file...
+  papplJobSetImpressions(job, 1);
+
+  if ((fd = open(papplJobGetFilename(job), O_RDONLY)) < 0)
+  {
+    papplLogJob(job, PAPPL_LOGLEVEL_ERROR, "Unable to open print file '%s': %s", papplJobGetFilename(job), strerror(errno));
+    return (false);
+  }
+
+  while ((bytes = read(fd, buffer, sizeof(buffer))) > 0)
+  {
+    if (papplDeviceWrite(device, buffer, (size_t)bytes) < 0)
+    {
+      papplLogJob(job, PAPPL_LOGLEVEL_ERROR, "Unable to send %d bytes to printer.", (int)bytes);
+      close(fd);
+      return (false);
+    }
+  }
+  close(fd);
+
+  papplJobSetImpressionsCompleted(job, 1);
+
+  return (true);
+}
+
+// 'Brf_generic_rendjob()' - End a job.
+
+static bool // O - `true` on success, `false` on failure
+brf_gen_rendjob(
+    pappl_job_t *job,            // I - Job
+    pappl_pr_options_t *options, // I - Job options
+    pappl_device_t *device)      // I - Output device
+{
+  (void)job;
+  (void)options;
+  (void)device;
+
+  return (true);
+}
+
+// 'Brf_generic_rendpage()' - End a page.
+
+static bool // O - `true` on success, `false` on failure
+brf_gen_rendpage(
+    pappl_job_t *job,            // I - Job
+    pappl_pr_options_t *options, // I - Job options
+    pappl_device_t *device,      // I - Output device
+    unsigned page)               // I - Page number
+{
+  (void)job;
+  (void)page;
+
+  papplDevicePuts(device, "P1\n");
+
+  return (true);
+}
+
+// 'Brf_generic_rstartjob()' - Start a job.
+
+static bool // O - `true` on success, `false` on failure
+brf_gen_rstartjob(
+    pappl_job_t *job,            // I - Job
+    pappl_pr_options_t *options, // I - Job options
+    pappl_device_t *device)      // I - Output device
+{
+  (void)job;
+  (void)options;
+  (void)device;
+
+  return (true);
+}
+
+// 'brf_gen_rwriteline()' - Write a raster line.
+
+static bool // O - `true` on success, `false` on failure
+brf_gen_rwriteline(
+    pappl_job_t *job,            // I - Job
+    pappl_pr_options_t *options, // I - Job options
+    pappl_device_t *device,      // I - Output device
+    unsigned y,                  // I - Line number
+    const unsigned char *line)   // I - Line
+{
+  if (line[0] || memcmp(line, line + 1, options->header.cupsBytesPerLine - 1))
+  {
+    unsigned i;                   // Looping var
+    const unsigned char *lineptr; // Pointer into line
+    unsigned char buffer[10000007],
+        *bufptr; // Pointer into buffer
+
+    for (i = options->header.cupsBytesPerLine, lineptr = line, bufptr = buffer; i > 0; i--)
+      *bufptr++ = ~*lineptr++;
+
+    papplDevicePrintf(device, "GW0,%u,%u,1\n", y, options->header.cupsBytesPerLine);
+    papplDeviceWrite(device, buffer, options->header.cupsBytesPerLine);
+    papplDevicePuts(device, "\n");
+  }
```

Comment:
`brf_gen_rwriteline` shouldn't be getting called, it'd mean that the filtering declarations went wrong. Better not do anything in this function but just log an error.